### PR TITLE
Split `flatten` from `techmap` and simplify it

### PR DESCRIPTION
**Pull request by whitequark**
Although the two passes started out very similar, they diverged over time and now have little in common. Moreover, `techmap` is extremely complex while `flatten` does not have to be, and this complexity interferes with improving `flatten`.

This is a fairly large PR, and it is best reviewed commit by commit: it starts by copying the techmap code wholesale and then simplifying it to make it easier to convince yourself that the changes are indeed correct. It can also be helpful to append `?w=1` to the URL to omit whitespace changes from the output.

The only actual improvement to `flatten` in this PR is using topological order, but it sets the stage for adding support for `hdlname` and processes (without `proc`).

### CODE REVIEW

**Comment by whitequark** [comment]
The "topologically sort modules" commit started out as a simplification (because I didn't fully understand the logic) and ended up as more of an optimization (but now the logic is documented!). I hope that's ok.

File: passes/techmap/flatten.cc [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+				if (tpl != nullptr) {
+					if (topo_modules.database.count(tpl) == 0)
+						worklist.insert(tpl);
+					topo_modules.edge(tpl, module);
+				}
+			}
+		}
+
+		if (!topo_modules.sort())
+			log_error("Cannot flatten a design that contains loops.\n");
```
Comment (daveshah1) [inline passes/techmap/flatten.cc:341]: This should probably be phrased better, `loops` to me means combinational loops rather than loops in the hierarchy

Comment (whitequark) [reply passes/techmap/flatten.cc:341]: Makes sense. Any suggestion for phrasing? `Cannot flatten a recursive design`?

Comment (daveshah1) [reply passes/techmap/flatten.cc:341]: `design containing recursive instantiations`?

**Comment by whitequark** [comment]
Thanks a lot!

**Pull request merged**